File: 2018-8-5/Knapsack.cpp
#include <cstdio>
#include <algorithm>
using namespace std;
int arr[2005][2005];
int w[2005], v[2005];
int algo(int n, int C){
	for(int i = 0;i < n;i++){
		for(int c = 0;c <= C;c++){
			if (i == 0){
				if (w[i] > c){
					arr[i][c] = 0;
				}
				else{
					arr[i][c] = v[i];
				}
			}
			else{
				if (w[i] > c){
					arr[i][c] = arr[i - 1][c];
				}
				else{
					arr[i][c] = max(arr[i - 1][c], arr[i - 1][c - w[i]] + v[i]);
				}
			}
		}
	}
	// for(int i = 0;i < n;i++){
		// for(int c = 0;c <= C;c++){
			// printf("%d\t", arr[i][c]);
		// }
		// printf("\n");
	// }
	return arr[n - 1][C];
}
int main(){
	int S, N;
	scanf("%d %d", &S, &N);
	for(int i = 0;i < N;i++){
		scanf("%d %d", &w[i], &v[i]);
	}
	printf("%d", algo(N, S));
	return 0;
}

File: 2018-8-5/Knapsack.py
#from pprint import pprint
def algo(w, v, n, C):
    arr = [[None] * (C + 1) for i in range(n)]
    for i in range(n):
        for c in range(0, C + 1):
            if i == 0:
                arr[i][c] = 0 if w[i] > c else v[i]
            else:
                if w[i] > c:
                    arr[i][c] = arr[i - 1][c]
                else:
                    arr[i][c] = max(arr[i - 1][c], arr[i - 1][c - w[i]] + v[i])
    #pprint(arr)
    return arr[-1][-1]
S, N = map(int, input().rstrip().split())
w = []
v = []
for i in range(N):
    inp = input().rstrip().split()
    w.append(int(inp[0]))
    v.append(int(inp[1]))
print(algo(w, v, N, S))
